device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java

```diff
@@ -495,18 +510,150 @@ public void close(IotHubConnectionStatusChangeReason reason, Throwable cause)
                 this.updateStatus(IotHubConnectionStatus.DISCONNECTED, reason, cause);
 
                 // Notify send thread to finish up so it doesn't survive this close
-                synchronized (this.sendThreadLock)
+                this.sendThreadSemaphore.release();
+
+                // Notify receive thread to finish up so it doesn't survive this close
+                this.receiveThreadSemaphore.release();
+
+                // Notify reconnect thread to finish up so it doesn't survive this close
+                this.reconnectThreadSemaphore.release();
+
+                log.debug("Client connection closed successfully");
+            }
+        }
+    }
+
+    // should only be called from IotHubReconnectTask
+    public void reconnect() throws InterruptedException
+    {
+        synchronized (this.reconnectionLock)
+        {
+            long reconnectionStartTimeMillis = 0;
+            int reconnectionAttempt = 0;
+            String deviceSessionToReconnect = null;
+
+            RetryPolicy retryPolicy; // retry policy to be used for connection level retry, not device session specific retry
+            if (isMultiplexing)
+            {
+                retryPolicy = multiplexingRetryPolicy;
+            }
+            else
+            {
+                retryPolicy = this.getDefaultConfig().getRetryPolicy();
+            }
```

Comment:
Does Java have the turnery operator?

```suggestion
                retryPolicy = isMultiplexing
                    ? multiplexingRetryPolicy
                    : this.getDefaultConfig().getRetryPolicy();
```

Reply:
It does, yeah. I'll change this to use it